### PR TITLE
[runtime] Use a separate implementation of mono-mmap.c for wasm, the default mmap based implementation doesn't work because munmap () cannot free memory inside mappings.

### PR DESCRIPTION
**Pull request by vargaz**

### CODE REVIEW

File: mono/utils/mono-mmap-wasm.c [hunk 1]
```diff
@@ -0,0 +1,223 @@
+
+#include <config.h>
+
+#ifdef HOST_WASM
+
+#include <sys/types.h>
+#if HAVE_SYS_STAT_H
```
Comment (jaykrell) [inline mono/utils/mono-mmap-wasm.c]: `#if` needed, or HOST_WASM determines it all?

File: mono/utils/mono-mmap-wasm.c [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+	if (!malloced_shared_area)
+		malloced_shared_area = mono_malloc_shared_area (getpid ());
+	/* get the pid here */
+	return malloced_shared_area;
+}
+
+void
+mono_shared_area_remove (void)
+{
+	if (malloced_shared_area)
```
Comment (jaykrell) [inline mono/utils/mono-mmap-wasm.c]: `if` not needed.

File: mono/utils/mono-mmap-wasm.c [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+	if (length == 0)
+		/* emscripten throws an exception on 0 length */
+		return NULL;
+
+	BEGIN_CRITICAL_SECTION;
+	ptr = mmap (0, length, prot, mflags, fd, offset);
+	END_CRITICAL_SECTION;
+	if (ptr == MAP_FAILED)
+		return NULL;
+	*ret_handle = (void*)length;
```
Comment (jaykrell) [inline mono/utils/mono-mmap-wasm.c]: Is "definite assignment" needed? i.e. assign 0 in the error path, or unconditionally  before doing anything? Or no matter?

**Comment by kjpou1** [comment]
I have tested this with @vargaz and is good to go.

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by vargaz** [comment]
@monojenkins build OS X i386

**Pull request merged**